Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp

```diff
@@ -294,27 +256,35 @@ Ref<CoordinatedTileBuffer> SkiaPaintingEngine::paintLayer(const GraphicsLayer& l
 
 Ref<CoordinatedTileBuffer> SkiaPaintingEngine::postPaintingTask(const GraphicsLayer& layer, RenderingMode renderingMode, const IntRect& dirtyRect, bool contentsOpaque, float contentsScale)
 {
-    WTFBeginSignpost(this, RecordTile);
-    auto displayList = recordDisplayList(renderingMode, layer, dirtyRect, contentsOpaque, contentsScale);
-    WTFEndSignpost(this, RecordTile);
-
     auto buffer = createBuffer(renderingMode, dirtyRect.size(), contentsOpaque);
     buffer->beginPainting();
 
+    WTFBeginSignpost(this, RecordTile);
+    auto* recordingCanvas = m_pictureRecorder.beginRecording(dirtyRect.width(), dirtyRect.height());
+    GraphicsContextSkia recordingContext(*recordingCanvas, renderingMode, RenderingPurpose::LayerBacking);
+    recordingContext.beginRecording();
+    paintIntoGraphicsContext(layer, recordingContext, dirtyRect, contentsOpaque, contentsScale);
+    auto imageToFenceMap = recordingContext.endRecording();
+    auto picture = m_pictureRecorder.finishRecordingAsPicture();
+    WTFEndSignpost(this, RecordTile);
+
     auto& workerPool = renderingMode == RenderingMode::Accelerated ? *m_gpuWorkerPool.get() : *m_cpuWorkerPool.get();
-    workerPool.postTask([buffer = Ref { buffer }, displayList = WTFMove(displayList), dirtyRect]() mutable {
+    workerPool.postTask([buffer = Ref { buffer }, picture = sk_sp<SkPicture> { picture }, dirtyRect, imageToFenceMap = WTFMove(imageToFenceMap)]() mutable {
         if (auto* canvas = buffer->canvas()) {
+            auto replayCanvas = SkiaReplayCanvas::create(dirtyRect.size(), WTFMove(imageToFenceMap));
+            replayCanvas->addCanvas(canvas);
```

Comment:
I think we should only use the SkiaReplayCanvas when doing GPU rendering. We could even avoid it if the image to fence map is empty, not sure it's worth it, though.

Reply:
Good moment. EWS is green as it-is, so we can now optimize and avoid the SkiaReplayCanvas creation altogether whenever imageFencesMap is empty -- in both CPU and GPU drawing scenarios, e.g. a basic fillRect operation won't need a SkiaReplayCanvas in GPU rendering mode either, so I think it's the cleanest approach to decide based on the image to fence map.